tests/remote_attestation/maa/sgxlkl_enclave/tlscli.c

```diff
@@ -0,0 +1,561 @@
+        ret = r;
+        goto done;
+    }
+
+    if ((r = mbedtls_ssl_setup(&cli->ssl, &cli->conf)) != 0)
+    {
+        _put_mbedtls_err(err, r, "mbedtls_ssl_setup");
+        ret = r;
+        goto done;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+int tlscli_connect(
+    bool debug,
+    const char* host,
+    const char* port,
+    const char* crt_path,
+    const char* pk_path,
+    tlscli_t** cli_out,
+    tlscli_err_t* err)
+{
+    int ret = -1;
+    int r;
+    tlscli_t* cli = NULL;
+
+    _clear_err(err);
+
+    if (cli_out)
+        *cli_out = NULL;
+
+    if (!_started)
+    {
+        _put_err(err, "not started: please call tlscli_startup()");
+        goto done;
+    }
+
+    if (!cli_out)
+    {
+        _put_err(err, "invalid cli parameter");
+        goto done;
+    }
+
+    if (!host)
+    {
+        _put_err(err, "invalid host parameter");
+        goto done;
+    }
+
+    if (!port)
+    {
+        _put_err(err, "invalid port parameter");
+        goto done;
+    }
+
+    /* Initialize the cli structure */
+    {
+        if (!(cli = calloc(1, sizeof(tlscli_t))))
+        {
+            _put_err(err, "calloc() failed: out of memory");
+            goto done;
+        }
+
+        mbedtls_net_init(&cli->net);
+        mbedtls_ssl_init(&cli->ssl);
+        mbedtls_ssl_config_init(&cli->conf);
+        mbedtls_x509_crt_init(&cli->crt);
+        mbedtls_pk_init(&cli->pk);
+    }
+
+    if ((r = mbedtls_net_connect(
+             &cli->net, host, port, MBEDTLS_NET_PROTO_TCP)) != 0)
+    {
+        _put_mbedtls_err(err, r, "mbedtls_net_connect()");
+        ret = r;
+        goto done;
+    }
+
+    if ((r = _configure_cli(cli, debug, crt_path, pk_path, err)) != 0)
+    {
+        ret = r;
+        goto done;
+    }
+
+    if ((r = mbedtls_ssl_set_hostname(&cli->ssl, host)) != 0)
+    {
+        _put_mbedtls_err(err, r, "mbedtls_ssl_set_hostname");
+        ret = r;
+        goto done;
+    }
+
+    mbedtls_ssl_set_bio(
+        &cli->ssl, &cli->net, mbedtls_net_send, mbedtls_net_recv, NULL);
+
+    while ((r = mbedtls_ssl_handshake(&cli->ssl)) != 0)
+    {
+        if (r != MBEDTLS_ERR_SSL_WANT_READ && r != MBEDTLS_ERR_SSL_WANT_WRITE)
+        {
+            _put_mbedtls_err(err, r, "mbedtls_ssl_handshake");
+            ret = r;
+            goto done;
+        }
+    }
+
+    if (mbedtls_ssl_get_verify_result(&cli->ssl) != 0)
+    {
+        mbedtls_ssl_close_notify(&cli->ssl);
+        _put_err(err, "handshake failed");
+        goto done;
+    }
+
+    *cli_out = cli;
+    cli = NULL;
+
+    ret = 0;
+
+done:
+
+    if (cli)
+    {
+        mbedtls_ssl_free(&cli->ssl);
+        mbedtls_net_free(&cli->net);
+        mbedtls_ssl_config_free(&cli->conf);
+        mbedtls_x509_crt_free(&cli->crt);
+        mbedtls_pk_free(&cli->pk);
+        free(cli);
+    }
+
+    return ret;
+}
+
+int tlscli_read(tlscli_t* cli, void* data, size_t size, tlscli_err_t* err)
```

Comment:
I don't think we need tlscli_write or tlscli_read

Reply:
Removed